File: backend/BeElite/src/main/java/com/salesianos/triana/BeElite/dto/User/AthleteDetailsDto.java
package com.salesianos.triana.BeElite.dto.User;

import com.salesianos.triana.BeElite.dto.Program.InviteDto;
import com.salesianos.triana.BeElite.dto.Program.ProgramDto;
import com.salesianos.triana.BeElite.model.Athlete;
import jakarta.validation.constraints.NotEmpty;
import lombok.Builder;

import java.util.List;
import java.util.Optional;

@Builder
public record AthleteDetailsDto(@NotEmpty(message = "Athlete must have username.")
                                String username,
                                @NotEmpty(message = "Athlete must have a name.")
                                String name,
                                String profilePicUrl,
                                @NotEmpty(message = "Athlete must have email.")
                                String email,
                                ProgramDto program,
                                UserDto coach,
                                int completed_sessions,
                                List<InviteDto> invites
                                )
{
    public static AthleteDetailsDto of(Athlete a){
        return AthleteDetailsDto.builder()
                .username(a.getUsername())
                .name(a.getName())
                .profilePicUrl(a.getProfilePicUrl())
                .email(a.getEmail())
                .program(a.getProgram() != null ? ProgramDto.of(a.getProgram()) : ProgramDto.empty())
                .coach(a.getProgram() != null ? UserDto.of(a.getProgram().getCoach()) : UserDto.empty())
                .completed_sessions(a.getCompleted_sessions())
                .invites(a.getInvites() != null ? a.getInvites().stream().map(InviteDto::of).toList() : InviteDto.emptyList())
                .build();
    }
}

File: frontend/be_elite/lib/models/Athlete/athlete_details_dto/athlete_details_dto.dart
import 'package:be_elite/models/Coach/user_dto.dart';
import 'package:be_elite/models/Program/invite_dto.dart';
import 'package:be_elite/models/Program/program_dto.dart';

class AthleteDetailsDto {
  String? username;
  String? name;
  String? profilePicUrl;
  String? email;
  ProgramDto? program;
  UserDto? coach;
  int? completedSessions;
  List<InviteDto>? invites;

  AthleteDetailsDto({
    this.username,
    this.name,
    this.profilePicUrl,
    this.email,
    this.program,
    this.coach,
    this.completedSessions,
    this.invites,
  });

  factory AthleteDetailsDto.fromJson(Map<String, dynamic> json) {
    return AthleteDetailsDto(
      username: json['username'] as String?,
      name: json['name'] as String?,
      profilePicUrl: json['profilePicUrl'] as String?,
      email: json['email'] as String?,
      program: json['program'] == null
          ? null
          : ProgramDto.fromJson(json['program'] as Map<String, dynamic>),
      coach: json['coach'] == null
          ? null
          : UserDto.fromJson(json['coach'] as Map<String, dynamic>),
      completedSessions: json['completed_sessions'] as int?,
      invites: (json['invites'] as List<dynamic>?)
          ?.map((e) => InviteDto.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }

  Map<String, dynamic> toJson() => {
        'username': username,
        'name': name,
        'profilePicUrl': profilePicUrl,
        'email': email,
        'program': program?.toJson(),
        'coach': coach?.toJson(),
        'completed_sessions': completedSessions,
        'invites': invites?.map((e) => e.toJson()).toList(),
      };
}
